File: iterator/python/implementations/physical_book_iterator.py
from typing import List, Optional
from interfaces.iterator import Iterator
from interfaces.book import Book

class PhysicalBookIterator(Iterator):
    def __init__(self, books: List[Book]):
        self._books = books
        self._position = 0

    def first(self) -> Optional[Book]:
        self._position = 0
        return self.current_item()

    def next(self) -> None:
        self._position += 1

    def is_done(self) -> bool:
        return self._position >= len(self._books)

    def current_item(self) -> Optional[Book]:
        return None if self.is_done() else self._books[self._position]

File: iterator/java/implementations/PhysicalBookIterator.java
package iterator.java.implementations;

import java.util.List;

import iterator.java.interfaces.Book;
import iterator.java.interfaces.Iterator;

public class PhysicalBookIterator implements Iterator {
    private List<Book> books;
    private int position = 0;

    public PhysicalBookIterator(List<Book> books) {
        this.books = books;
    }

    @Override
    public Book first() {
        position = 0;
        return currentItem();
    }

    @Override
    public void next() {
        position++;
    }

    @Override
    public boolean isDone() {
        return position >= books.size();
    }

    @Override
    public Book currentItem() {
        return isDone() ? null : books.get(position);
    }
}